Milestone 5 - Request for code review

javitocor opened the pull request:
In this milestone we have added the following features:
- Added friendship model
- Added associations between user and friendship models
- Added new routes to send a friendship request and accept or decline friendship
- Added link to request friendship in users#index and user#show views
- Added links to accept or decline the friendship in user's show page
- Added helper method to check if users are already friends
- Added helper method to check if there is a friendship request pending for a user
- Added integration and unit tests with rspec, capybara and shoulda for all the features and methods

javitocor committed:
Modified routes.rb

javitocor committed:
Generated rspec install

javitocor committed:
Set up rspec and capybara config

javitocor committed:
Added friendship model

javitocor committed:
Added associations in user and friendship models

javitocor committed:
Added friendship actions to users controllers

tirthajyoti-ghosh committed:
added send friendship request functionality

javitocor committed:
Added friends check helper method and updated index and show users views

javitocor committed:
Added confirm and decline methods to friendships

javitocor committed:
Added logic to show and index users views, added pending_request helper

javitocor committed:
Created files for tests

javitocor committed:
Added tests to users controller and models

javitocor committed:
Added tests for application helper

javitocor committed:
Fixed rubocop

javitocor committed:
Modified rubocop.yml

tirthajyoti-ghosh committed:
changed version of chromedriver-helper gem

majovanilla commented on app/controllers/users_controller.rb:
```diff
@@ -8,5 +8,24 @@ def index
+    @pending_requests = @user.inverse_friendships.where(confirmed: false)
+  end
+
+  def send_friend
+    Friendship.create(user_id: current_user.id, friend_id: params[:id], confirmed: false)
+
+    redirect_to users_path, notice: 'Friend request sent.'
+  end
```
Right now, there is no place on your app where a user can see the friend requests they sent or received (on the user's profile as you have it, it's super difficult to find). Please implement this menu (could be called Notifications).

Also, when a user sends a friendship request, the other user sees "Friendship already sent" in the All Users menu, instead of the accept or decline button. Please fix this.

majovanilla commented on app/controllers/users_controller.rb:
```diff
@@ -8,5 +8,24 @@ def index
+    @friendship = Friendship.find(params[:id])
+    @friendship.destroy
+    redirect_to current_user
   end
```
You should also add the option on the view to reject the friendship if the user doesn't want to be friends.

majovanilla commented on app/controllers/users_controller.rb:
```diff
@@ -8,5 +8,24 @@ def index
+    @friendship = Friendship.find(params[:id])
+    @friendship.update(confirmed: true)
+    redirect_to current_user
+  end
```
You should add a button to accept the friendship.

majovanilla commented on app/views/users/index.html.erb:
```diff
@@ -4,8 +4,15 @@
-          <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
+          <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %> | 
         </span>
+        <% if current_user != user && friends_check(user) == false && pending_request(user) == false %>
+          <%= link_to 'Send friend request', "/users/#{user.id}/send_friend", class: 'profile-link' %>
```
This should be a button. Please adjust the styling.

majovanilla commented on app/views/users/show.html.erb:
```diff
@@ -1,5 +1,23 @@
+          <li><%= request.user.name %> | <%= accept_friend(request) %> | <%= reject_friend(request) %></li>
+        <% end %>
+      </ul>
+    <% end %>
```
There is something wrong with this part. When the user receives an invitation, I see this message in the profile, instead of the accept / reject buttons. It's not clear who I am adding as a friend and it says send friend request instead of accept it!
![Screenshot from 2020-05-27 13-42-16](https://user-images.githubusercontent.com/51926937/83059758-087c8f80-a020-11ea-9181-1fb273477eea.png)

majovanilla commented on app/views/users/show.html.erb:
```diff
@@ -1,5 +1,23 @@
+    <% end %>
+  <% end %>
+
   <article class="timeline">
     <h3>Recent posts:</h3>
     <ul class="posts">
```
A non-friend shouldn't be able to see the posts of a user when they check their profile. You should only have the send friend request button.

That should also happen in the timeline: only show the current user's and confirmed friend's posts.

majovanilla commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -0,0 +1,30 @@
+  context 'Test decline friend method' do
+    it 'should return a decline link' do
+      expect(reject_friend(friendship)).to have_content('Decline')
+    end
```
What exactly are you testing here? You check the method but you don't check the view. If you don't use the method on your view, this check is not actually doing anything.

majovanilla commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    click_link 'Decline'
+    visit users_path
+    expect(page).to have_content('Send friend request')
+  end
```
You have the tests but in the real app you don't have those links. Please make sure you are actually testing what you want to test.

majovanilla reviewed (changes requested):
# Code Review in progress 👀 by Maria Reyes


Hey team, 👋
## You did a *great* job on this project! 💜 
- Good Readme template
- Good Pull Request Message

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Improve the project description on your Readme and add some information on how your app works and what it can do. 

- Please improve the organization of your tests. You should separate the unit and integration tests in their own folders and always keep one `describe` in each file. That way, someone who doesn't know your project can easily find what they are looking for! This will be of a lot of help when you work in larger, collaborative projects.

- Currently, you have some missing requirements (I leave a summary here, check details on the code below) :
   - User can see friendships requests 
   - User has to accept a friendship request to start a friendship
   - User can reject a friendship request
   - There is a button to invite user to friendship on users list page and on single user page
   - Posts timeline includes posts written by logged-in user and their friends

### Optional Changes:

If you want to make your project even better 🔥💥, I have a few suggestions you can implement.  
These are completely [Optional]: 🦔

- You don't have a link to access the current user profile. This would improve the UX a lot since right now it's quite confusing to see in which profile I am.

- Instead of having to create multiple users manually, you should use your seeds file (in the DB folder) and include some fake users to play with) See this [link](https://www.xyzpub.com/en/ruby-on-rails/3.2/seed_rb.html).

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

javitocor committed:
Added unfriend features and logic to users show view

javitocor committed:
Modified gemfile

javitocor committed:
Added link to current user in navbar

javitocor committed:
Added link to current user in navbar

javitocor committed:
Fixed show page error

javitocor committed:
Added faker gem

javitocor committed:
Created users and posts in db seeds

javitocor committed:
Modified readme file

javitocor committed:
Modified user controller rspec

javitocor committed:
Added methods to see only friends posts in timeline

javitocor committed:
Added accept and decline friendship buttons to users index

javitocor committed:
Added accept and decline friendship buttons to users profile

javitocor committed:
Redirect Update and destroy actions to same page

javitocor committed:
Modified users controller tests and added unfriend test

javitocor committed:
Added tests to posts controller

javitocor committed:
Added test to posts controllers

javitocor committed:
Fixed rubocop errors

javitocor commented:
We have applied the following changes to the project:
- We have added a friends list to the user profile
- We have added the unfriend button to all users index and users show views
- We have added a method to only display the current user and friends posts in the timeline
- We have added logic to only see posts of friends in the profiles
- We have added the accept/decline buttons to the index users view
- We have modified the readme file
- Now you can create 15 new users with rails db:seed
- We have added tests for posts controller
- We have modified and added tests for users controller
- We have fixed all rubocop errors

javitocor commented:
To clarify, the accept and decline buttons only appear when the receiver is the current user, you cannot see the friend request you have sent, just like on Facebook, you only will have the message 'Friend request pending'.

nikasulo reviewed (approved):
# Great work! @javitocor and @tirthajyoti-ghosh :medal_military: 
You have done really well building this. Please merge and keep moving forward, I wish you the very best and see you at the top!

##### Think I missed something? Need some help with this or any of your projects? Please feel free to reach out to me on [Twitter](https://twitter.com/oluwadamilareo_) or [Instagram](https://instagram.com/oluwadamilare_olusakin), even if you just want to chat and I will respond to you as soon as I can.
